nameko/dependencies.py

```diff
@@ -3,64 +3,65 @@
 """
 from __future__ import absolute_import
 
-from functools import wraps, partial
+from functools import partial
 import inspect
-from itertools import chain
 import types
-from weakref import WeakSet, WeakKeyDictionary
+import weakref
 
 from eventlet.event import Event
+
 from nameko.utils import SpawningSet, repr_safe_str
 
 from logging import getLogger
 _log = getLogger(__name__)
 
-ENTRYPOINT_PROVIDERS_ATTR = 'nameko_entrypoints'
-
 
-# constants for dependency sharing
-CONTAINER_SHARED = object()
-# PROCESS_SHARED also serves as weakref-able sentinel obj
-PROCESS_SHARED = type('process', (), {})()
+ENTRYPOINT_EXTENSIONS_ATTR = 'nameko_entrypoints'
+UNBOUND_NAME = "<unbound-extension>"
 
+shared_extensions = weakref.WeakKeyDictionary()
 
-class NotInitializedError(Exception):
-    pass
 
+class Extension(object):
 
-class DependencyTypeError(TypeError):
-    pass
+    __state = None
+    __shared = False
+    __clone = False
 
+    name = UNBOUND_NAME  # property, knows if clone?
+    container = None
 
-class DependencyProvider(object):
-
-    bound = False
-    name = "<unbound-provider>"
+    def __init__(self, *args, **kwargs):
+        # sharing status is not persisted into state;
+        # cloned extensions will never be shared
+        self.__shared = kwargs.pop('shared', False)
```

Comment:
i though this was going to happen via different `bind` implementations rather than params